src/hotspot/share/classfile/systemDictionary.cpp

```diff
@@ -496,53 +512,54 @@ void SystemDictionary::double_lock_wait(JavaThread* thread, Handle lockObject) {
                                                           protection_domain,
                                                           true,
-                                                          CHECK_NULL);
-
-  // parallelCapable class loaders do NOT wait for parallel superclass loads to complete
-  // Serial class loaders and bootstrap classloader do wait for superclass loads
- if (!class_loader.is_null() && is_parallelCapable(class_loader)) {
-    MutexLocker mu(THREAD, SystemDictionary_lock);
-    return dictionary->find_class(name_hash, name);
-  }
-
-  // must loop to both handle other placeholder updates
-  // and spurious notifications
-  bool super_load_in_progress = true;
-  PlaceholderEntry* placeholder;
-  while (super_load_in_progress) {
-    MutexLocker mu(THREAD, SystemDictionary_lock);
-    // Check if classloading completed while we were loading superclass or waiting
-    InstanceKlass* check = dictionary->find_class(name_hash, name);
-    if (check != NULL) {
-      // Klass is already loaded, so just return it
-      return check;
+                                                          CHECK);
+}
+
+// parallelCapable class loaders do NOT wait for parallel superclass loads to complete
+// Serial class loaders and bootstrap classloader do wait for superclass loads
+static bool should_wait_for_loading(Handle class_loader) {
+  return class_loader.is_null() || !is_parallelCapable(class_loader);
+}
+
+// For bootstrap and non-parallelCapable class loaders, check and wait for
+// another thread to complete loading this class.
+InstanceKlass* SystemDictionary::handle_parallel_loading(JavaThread* current,
+                                                         unsigned int name_hash,
+                                                         Symbol* name,
+                                                         ClassLoaderData* loader_data,
+                                                         Handle lockObject,
+                                                         bool* throw_circularity_error) {
+  PlaceholderEntry* oldprobe = placeholders()->get_entry(name_hash, name, loader_data);
+  if (oldprobe != NULL) {
+    // only need check_seen_thread once, not on each loop
+    // 6341374 java/lang/Instrument with -Xcomp
+    if (oldprobe->check_seen_thread(current, PlaceholderTable::LOAD_INSTANCE)) {
+      log_circularity_error(oldprobe);
+      *throw_circularity_error = true;
+      return NULL;
     } else {
-      placeholder = placeholders()->get_entry(name_hash, name, loader_data);
-      if (placeholder != NULL && placeholder->super_load_in_progress()) {
+      // Wait until the first thread has finished loading this class. Also wait until all the
+      // threads trying to load its super class have removed their placeholders.
+      while (oldprobe != NULL &&
+             (oldprobe->instance_load_in_progress() || oldprobe->super_load_in_progress())) {
```

Comment:
This loop in handle_parallel_super_load and the one in resolve_instance_class_or_null are the same loop.  The current code waits for all the LOAD_SUPER placeholders to be removed.  If there are 'n' threads loading a class with a super, the first thread will add a LOAD_INSTANCE placeholder, and the second thread finding LOAD_INSTANCE will wait at the loop in resolve_instance_class_or_null for the LOAD_INSTANCE to be done.
If the first thread loads the class, and calls resolve_super_or_fail and adds the LOAD_SUPER placeholder, the next threads will call handle_parallel_super_load and wait for all the LOAD_SUPER placeholder to be removed from all the threads calling handle_parallel_super_load.
Once the first thread completes loading the super class, and removes the LOAD_SUPER placeholders, and all the other threads complete trying to load the super class, the other threads will fall into the loop in resolve_instance_class_or_null, waiting for the first thread to remove LOAD_INSTANCE.
There's only one LOAD_INSTANCE placeholder for each class/class loader pair in progress. There may be many LOAD_SUPER placeholders. The point is that both conditions can be checked in the same loop.  The loop needs to wait for all LOAD_SUPERs and the main LOAD_INSTANCE to clear.
I could rename "handle_parallel_loading" to "wait_for_parallel_loading" if that makes it more clear(?)